std/digest/xxh.d

```diff
@@ -0,0 +1,3352 @@
+    0x95, 0x16, 0x04, 0x28, 0xaf, 0xd7, 0xfb, 0xca, 0xbb, 0x4b, 0x40, 0x7e,
+];
+
+/*
+ * Downcast + upcast is usually better than masking on older compilers like
+ * GCC 4.2 (especially 32-bit ones), all without affecting newer compilers.
+ *
+ * The other method, (x & 0xFFFFFFFF) * (y & 0xFFFFFFFF), will AND both operands
+ * and perform a full 64x64 multiply -- entirely redundant on 32-bit.
+ */
+private xxh_u64 XXH_mult32to64(xxh_u32 x, xxh_u32 y) @trusted pure nothrow @nogc
+{
+    return (cast(xxh_u64)(x) * cast(xxh_u64)(y));
+}
+
+/*!
+ * @brief Calculates a 64.128-bit long multiply.
+ *
+ * Uses `__uint128_t` and `_umul128` if available, otherwise uses a scalar
+ * version.
+ *
+ * @param lhs , rhs The 64-bit integers to be multiplied
+ * @return The 128-bit result represented in an @ref XXH128_hash_t.
+ */
+private XXH128_hash_t XXH_mult64to128(xxh_u64 lhs, xxh_u64 rhs) @trusted pure nothrow @nogc
+{
+    static if (is(ucent))
+    {
+        //import std.int128;
+        //alias uint128_t = Int128;
+
+        const uint128_t product = cast(uint128_t_) lhs * cast(uint128_t_) rhs;
+        XXH128_hash_t r128;
+        // r128.low64  = cast(xxh_u64) (product);
+        // r128.high64 = cast(xxh_u64) (product >> 64);
+        r128.low64 = product.data.lo;
+        r128.high64 = product.data.hi;
+    }
+    else
+    {
+        /* First calculate all of the cross products. */
+        const xxh_u64 lo_lo = XXH_mult32to64(lhs & 0xFFFFFFFF, rhs & 0xFFFFFFFF);
+        const xxh_u64 hi_lo = XXH_mult32to64(lhs >> 32, rhs & 0xFFFFFFFF);
+        const xxh_u64 lo_hi = XXH_mult32to64(lhs & 0xFFFFFFFF, rhs >> 32);
+        const xxh_u64 hi_hi = XXH_mult32to64(lhs >> 32, rhs >> 32);
+
+        /* Now add the products together. These will never overflow. */
+        const xxh_u64 cross = (lo_lo >> 32) + (hi_lo & 0xFFFFFFFF) + lo_hi;
+        const xxh_u64 upper = (hi_lo >> 32) + (cross >> 32) + hi_hi;
+        const xxh_u64 lower = (cross << 32) | (lo_lo & 0xFFFFFFFF);
+
+        XXH128_hash_t r128;
+        r128.low64 = lower;
+        r128.high64 = upper;
+
+    }
+    return r128;
+}
+
+/*!
+ * @brief Calculates a 64-bit to 128-bit multiply, then XOR folds it.
+ *
+ * The reason for the separate function is to prevent passing too many structs
+ * around by value. This will hopefully inline the multiply, but we don't force it.
+ *
+ * @param lhs , rhs The 64-bit integers to multiply
+ * @return The low 64 bits of the product XOR'd by the high 64 bits.
+ * @see XXH_mult64to128()
+ */
+private xxh_u64 XXH3_mul128_fold64(xxh_u64 lhs, xxh_u64 rhs) @trusted pure nothrow @nogc
+{
+    XXH128_hash_t product = XXH_mult64to128(lhs, rhs);
+    return product.low64 ^ product.high64;
+}
+
+/*! Seems to produce slightly better code on GCC for some reason. */
+private xxh_u64 XXH_xorshift64(xxh_u64 v64, int shift) @trusted pure nothrow @nogc
+{
+    assert(0 <= shift && shift < 64, "shift out of range");
```

Comment:
You can use contract programming style.

Reply:
Yes, just another great D feature ;-)

Reply:
@ljmf00  Is the use of contract programming mandatory? I just gave it some try, but I reduces readabilty and bloats code.

Reply:
I will retry my current changes with `in(..)` instead of `in { .. }`. This might be much more readable.